malariagen_data/anopheles.py

```diff
@@ -5012,6 +5118,330 @@ def plot_g123_calibration(
+
+        except CacheMiss:
+            results = self._xpehh_gwss(**params)  # self.
+            self.results_cache_set(name=name, params=params, results=results)
+
+        x = results["x"]
+        xpehh = results["xpehh"]
+
+        return x, xpehh
+
+    def _xpehh_gwss(
+        self,
+        *,
+        contig,
+        analysis,
+        sample_sets,
+        cohort1_query,
+        cohort2_query,
+        window_size,
+        percentiles,
+        standardize,
+        standardization_bins,
+        standardization_n_bins,
+        standardization_diagnostics,
+        filter_min_maf,
+        map_pos,
+        min_ehh,
+        max_gap,
+        gap_scale,
+        include_edges,
+        use_threads,
+        min_cohort_size,
+        max_cohort_size,
+        random_seed,
+    ):
+        ds_haps1 = self.haplotypes(
+            region=contig,
+            analysis=analysis,
+            sample_query=cohort1_query,
+            sample_sets=sample_sets,
+            min_cohort_size=min_cohort_size,
+            max_cohort_size=max_cohort_size,
+            random_seed=random_seed,
+        )
+
+        ds_haps2 = self.haplotypes(
+            region=contig,
+            analysis=analysis,
+            sample_query=cohort2_query,
+            sample_sets=sample_sets,
+            min_cohort_size=min_cohort_size,
+            max_cohort_size=max_cohort_size,
+            random_seed=random_seed,
+        )
+
+        gt1 = allel.GenotypeDaskArray(ds_haps1["call_genotype"].data)
+        gt2 = allel.GenotypeDaskArray(ds_haps2["call_genotype"].data)
+        with self._dask_progress(desc="Load haplotypes"):
+            ht1 = gt1.to_haplotypes().compute()
+            ht2 = gt2.to_haplotypes().compute()
+
+        ac1 = ht1.count_alleles(max_allele=1)
+        ac2 = ht2.count_alleles(max_allele=1)
+        pos = ds_haps1["variant_position"].values
+
+        # To be modified to take both into account
+        if filter_min_maf > 0:
+            ac = ac1 + ac2
+            af = ac.to_frequencies()
+            maf = np.min(af, axis=1)
+            maf_filter = maf > filter_min_maf
+
+            ht1 = ht1.compress(maf_filter, axis=0)
+            ht2 = ht2.compress(maf_filter, axis=0)
+            pos = pos[maf_filter]
+            ac1 = ac1[maf_filter]
+            ac2 = ac2[maf_filter]
+
+        # compute XP-EHH
+        xp = allel.xpehh(
+            h1=ht1,
+            h2=ht2,
+            pos=pos,
+            map_pos=map_pos,
+            # min_maf=compute_min_maf,
+            min_ehh=min_ehh,
+            include_edges=include_edges,
+            max_gap=max_gap,
+            gap_scale=gap_scale,
+            use_threads=use_threads,
+        )
+
+        # remove any NaNs
+        na_mask = ~np.isnan(xp)
+        xp = xp[na_mask]
+        pos = pos[na_mask]
+        ac1 = ac1[na_mask]
+        ac2 = ac2[na_mask]
+
+        # take absolute value
+        xp = np.fabs(xp)
```

Comment:
Suggest we do not take the absolute value, because the sign is informative about which population has the longer haplotypes and therefore is experiencing selection.